drivers/rtc/rtc_counter.c

```diff
@@ -0,0 +1,464 @@
+
+	/* -EINVAL on overflow/invalid time */
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Stop counter */
+	ret = counter_stop(config->counter_dev);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = counter_get_value(config->counter_dev, &now_ticks);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Update the software offset: offset = desired_time - current_raw_ticks */
+	K_SPINLOCK(&data->lock) {
+		data->epoch_offset = (int64_t)desired_ticks - (int64_t)now_ticks;
+	}
+
+	/* Restart counter */
+	return counter_start(config->counter_dev);
+}
+
+static int counter_rtc_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	const struct counter_rtc_config *config = dev->config;
+	struct counter_rtc_data *data = dev->data;
+	uint32_t now_ticks;
+	int ret;
+	int64_t epoch;
+	int64_t current_seconds;
+
+	if (timeptr == NULL) {
+		return -EINVAL;
+	}
+
+	ret = counter_get_value(config->counter_dev, &now_ticks);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	K_SPINLOCK(&data->lock) {
+		epoch = data->epoch_offset;
+	}
+
+	current_seconds = (int64_t)now_ticks + epoch;
+
+	if (current_seconds < 0 || current_seconds > UINT32_MAX) {
+		return -ERANGE;
+	}
+
+	memset(timeptr, 0, sizeof(struct rtc_time));
+	counter_rtc_ticks_to_time((uint32_t)current_seconds, timeptr);
+
+	return 0;
+}
+
+#if defined(CONFIG_RTC_ALARM)
+
+static int counter_rtc_alarm_get_supported_fields(const struct device *dev, uint16_t id,
+						  uint16_t *mask)
+{
+	struct counter_rtc_data *data = dev->data;
+
+	if (!data->alarm_capable) {
+		return -ENOTSUP;
+	}
+
+	if (id >= data->num_alarm_chans) {
+		return -EINVAL;
+	}
+
+	*mask = (RTC_ALARM_TIME_MASK_SECOND | RTC_ALARM_TIME_MASK_MINUTE |
+		 RTC_ALARM_TIME_MASK_HOUR | RTC_ALARM_TIME_MASK_MONTHDAY |
+		 RTC_ALARM_TIME_MASK_MONTH | RTC_ALARM_TIME_MASK_YEAR |
+		 RTC_ALARM_TIME_MASK_WEEKDAY | RTC_ALARM_TIME_MASK_YEARDAY);
+
+	return 0;
+}
+
+static void counter_rtc_alarm_callback(const struct device *counter_dev, uint8_t chan_id,
+				       uint32_t ticks, void *user_data)
+{
+	struct counter_rtc_data *data = (struct counter_rtc_data *)user_data;
+	const struct device *rtc_dev = data->rtc_dev;
+
+	if (chan_id < data->num_alarm_chans && data->alarm_callback[chan_id]) {
+		data->alarm_callback[chan_id](rtc_dev, chan_id, data->alarm_user_data[chan_id]);
+		data->alarm_pending[chan_id] = false;
+	} else if (chan_id < data->num_alarm_chans) {
+		data->alarm_pending[chan_id] = true;
+	}
+}
+
+static int counter_rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+				      const struct rtc_time *timeptr)
+{
+	const struct counter_rtc_config *config = dev->config;
+	struct counter_rtc_data *data = dev->data;
+	uint32_t desired_ticks;
+	int ret;
+	int64_t epoch;
+	int64_t raw_alarm_ticks_64;
+	uint32_t top;
+	uint32_t alarm_ticks;
+	uint32_t now_raw;
+	struct counter_alarm_cfg alarm_cfg;
+
+	if ((mask && (timeptr == 0)) || (timeptr && !rtc_utils_validate_rtc_time(timeptr, mask))) {
```

Comment:
More implicit boolean conversions

```suggestion
	if ((mask && (timeptr == NULL)) || (timeptr != NULL && !rtc_utils_validate_rtc_time(timeptr, mask))) {
```